src/main/java/org/wildfly/security/auth/client/SetOAuth2ClientCredentialsConfiguration.java

```diff
@@ -0,0 +1,135 @@
+import static org.wildfly.security._private.ElytronMessages.log;
+import static org.wildfly.security.util.HttpClient.HttpHeader.authorizationBasic;
+
+import javax.json.Json;
+import javax.json.JsonObject;
+import javax.net.ssl.SSLContext;
+import javax.security.auth.callback.Callback;
+import javax.security.auth.callback.UnsupportedCallbackException;
+import java.io.IOException;
+import java.net.URI;
+import java.security.AccessController;
+import java.security.GeneralSecurityException;
+
+import org.wildfly.security.auth.callback.CredentialCallback;
+import org.wildfly.security.credential.BearerTokenCredential;
+import org.wildfly.security.util.HttpClient;
+import org.wildfly.security.util.HttpClient.HttpRequest;
+
+/**
+ * <p>A {@link AuthenticationConfiguration} that is capable of obtaining a {@link BearerTokenCredential} using
+ * the OAuth2 Client Credentials Grant Type.
+ *
+ * @author <a href="mailto:psilva@redhat.com">Pedro Igor</a>
+ */
+public class SetOAuth2ClientCredentialsConfiguration extends AuthenticationConfiguration {
+
+    private final URI tokenEndpointUri;
+    private final String clientId;
+    private final String clientSecret;
+    private final String scopes;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param tokenEndpointUri the OAuth2 Token Endpoint {@link URI}
+     * @param clientId         the client id
+     * @param clientSecret     the client secret
+     * @param scopes           a string with the scope of the access request
+     */
+    SetOAuth2ClientCredentialsConfiguration(URI tokenEndpointUri, String clientId, String clientSecret, String scopes) {
+        this.tokenEndpointUri = checkNotNullParam("tokenEndpointUri", tokenEndpointUri);
+        this.clientId = checkNotNullParam("clientId", clientId);
+        this.clientSecret = checkNotNullParam("clientSecret", clientSecret);
+        this.scopes = scopes;
+    }
+
+    /**
+     * Creates a new instance with a parent config.
+     *
+     * @param parent the parent config
+     * @param tokenEndpointUri the OAuth2 Token Endpoint {@link URI}
+     * @param clientId         the client id
+     * @param clientSecret     the client secret
+     * @param scopes           a string with the scope of the access request
+     */
+    SetOAuth2ClientCredentialsConfiguration(AuthenticationConfiguration parent, URI tokenEndpointUri, String clientId, String clientSecret, String scopes) {
+        super(parent);
+        this.tokenEndpointUri = checkNotNullParam("tokenEndpointUri", tokenEndpointUri);
+        this.clientId = checkNotNullParam("clientId", clientId);
+        this.clientSecret = checkNotNullParam("clientSecret", clientSecret);
+        this.scopes = scopes;
+    }
+
+    @Override
+    void handleCallbacks(AuthenticationConfiguration config, Callback[] callbacks) throws IOException, UnsupportedCallbackException {
+        for (Callback callback : callbacks) {
+            if (callback instanceof CredentialCallback) {

```

Comment:
Actually have a similar issue in Peter's PR - I think handleCallbacks should actually be able to go all the way to the top of the configuration so that the handleCallback method will be used instead - this looks like you will intercept all attempts at CallbackHandling even if something else can handle it.